rust/kernel/align.rs

```diff
@@ -0,0 +1,216 @@
+#[inline]
+pub fn ptr_align<T>(ptr: *const T, alignment: usize) -> *const T {
+    let ptr_val: usize = ptr as usize;
+    align(ptr_val, alignment) as *const T
+}
+
+/// Return downward aligned pointer of `ptr` to `T` according to the `alignment`
+/// requirement.
+#[allow(dead_code)]
+#[inline]
+pub fn ptr_align_down<T>(ptr: *const T, alignment: usize) -> *const T {
+    let ptr_val: usize = ptr as usize;
+    align_down(ptr_val, alignment) as *const T
+}
+
+/// Return the alignment of `val` according to `mask`.
+///
+/// `mask` must have continuous '1' which are little-end aligned.
+///
+/// This function is const and can be used in const context.
+#[allow(dead_code)]
+#[inline]
+pub const fn mask(val: usize, mask: usize) -> usize {
+    (val + mask) & !(mask)
+}
+
+/// Return the alignment of `alignment` bytes of `val`.
+///
+/// `alignment` must be a power of 2, and this function doesn't verify the
+/// requirement. It is the user's responsibility to take charge of this.
+///
+/// This function is const and can be used in const context.
+#[allow(dead_code)]
+#[inline]
+pub const fn align(val: usize, alignment: usize) -> usize {
+    let mask_val: usize = alignment - 1;
+    mask(val, mask_val)
+}
+
+/// Return the alignment downward of `alignment` bytes of `val`.
+///
+/// `alignment` must be a power of 2, and this function doesn't verify the
+/// requirement. It is the user's responsibility to take charge of this.
+///
+/// This function is const and can be used in const context.
+#[allow(dead_code)]
+#[inline]
+pub const fn align_down(val: usize, alignment: usize) -> usize {
+    let val: usize = val - (alignment - 1);
+    align(val, alignment)
+}
+
+/// Test if `val` is `alignment` bytes aligned.
+///
+/// `alignment` must be a power of 2, and this function doesn't verify the
+/// requirement. It is the user's responsibility to take charge of this.
+///
+/// This function is const and can be used in const context.
+#[allow(dead_code)]
+#[inline]
+pub const fn is_aligned(val: usize, alignment: usize) -> bool {
+    (val & (alignment - 1)) == 0
+}
+
+/// An interface for dealing with alignment.
+trait Align {
+    /// The type of alignment value. This type should be integer.
+    /// TODO: Bound this type to integer only.
+    type Alignment;
+
+    /// Alignment by mask
+    fn mask(&self, mask: Self::Alignment) -> Self;
+
+    /// Alignment by bytes.
+    fn align(&self, alignment: Self::Alignment) -> Self;
+
+    /// Alignment downward by bytes.
+    fn align_down(&self, alignment: Self::Alignment) -> Self;
+
+    /// Test if Aligned.
+    fn is_aligned(&self, alignment: Self::Alignment) -> bool;
+}
+
+/// A Helper macro for implementing `Align` trait for primitive integer types.
+macro_rules! impl_align_for_integer {
+    ($type: ty) => {
+        impl Align for $type {
+            type Alignment = Self;
+
+            #[inline]
+            fn mask(&self, mask: Self) -> Self {
+                (self + mask) & !(mask)
```

Comment:
I am not sure if it is proper to use the `const` function directly here. The const functions take `unsize` arguments, but for tait methods, I hope they can have their own `alignment` type.

For example, for `u64`, I hope the `mask` is `u64` as well. But if we use `const mask()` directly, it converts `u64` to `usize` implicitly.